go_util/util.go

```diff
@@ -31,3 +36,167 @@ func AppendPendingMetadata(testName string, link shared.MetadataLink, metadata *
+
+	// Open the META.yml file that we would write to, if it exists. In the case
+	// where it does exist we don't want to overwrite the file, so open in
+	// read-write-or-create mode.
+	filepath := dirpath + "/META.yml"
+	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if fi.Size() == 0 {
+		// This is an entirely new META.yml file, so create a new
+		// shared.Metadata representation for our expectation and write it out.
+		fmt.Println("CREATED-META-YML: for test", testName, "with url", link.URL)
+		metadata := shared.Metadata{
+			Links: []shared.MetadataLink{link},
+		}
+		writeToFile(metadata, f)
+		return
+	}
+
+	data, err := ioutil.ReadAll(f)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var metadata shared.Metadata
+	err = yaml.Unmarshal(data, &metadata)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	appendMetadataLink(testName, link, &metadata)
+	writeToFile(metadata, f)
+}
+
+// deleteMetadata deletes testname from metadata and return the new metadata and deleted metadataLink entries.
+func deleteMetadata(testname string, metadata shared.Metadata) (shared.Metadata, shared.MetadataLinks) {
+	deletedMetadataLinks := shared.MetadataLinks{}
+	newMetadataLinks := []shared.MetadataLink{}
+	for _, link := range metadata.Links {
+		deletedResults := []shared.MetadataTestResult{}
+		newResults := []shared.MetadataTestResult{}
+		for _, result := range link.Results {
+			if result.TestPath == testname {
+				// Add the deleted result to the metadataLinks
+				deletedResults = append(deletedResults, result)
+			} else {
+				newResults = append(newResults, result)
+			}
+		}
+		if len(deletedResults) > 0 {
+			deletedLink := shared.MetadataLink{Product: link.Product, URL: link.URL, Results: deletedResults}
+			deletedMetadataLinks = append(deletedMetadataLinks, deletedLink)
+		}
+
+		if len(newResults) > 0 {
+			newLink := shared.MetadataLink{Product: link.Product, URL: link.URL, Results: newResults}
+			newMetadataLinks = append(newMetadataLinks, newLink)
+		}
+	}
+	newMetadata := shared.Metadata{Links: newMetadataLinks}
+	return newMetadata, deletedMetadataLinks
+}
+
+// DeleteTestFromMetadata deletes a WPT testPath and returns the deleted metadata entries, if exists.
+func DeleteTestFromMetadata(testPath string) shared.MetadataLinks {
+	metadata := readMetadataFile(testPath)
+	if metadata == nil {
+		return shared.MetadataLinks{}
+	}
+
+	testname := path.Base(testPath)
+	newMetadata, metadataLinks := deleteMetadata(testname, *metadata)
+	// Nothing is deleted.
+	if len(metadataLinks) == 0 {
+		return shared.MetadataLinks{}
+	}
+
+	fmt.Println("Removing", testPath, "from the wpt-metadata repo")
+	// Everything is deleted
+	if len(newMetadata.Links) == 0 {
+		removeMetadataFile(testPath)
+	} else {
+		filepath := getYMLFilePath(testPath)
+		f, err := os.OpenFile(filepath, os.O_WRONLY|os.O_TRUNC, 0644)
```

Comment:
@foolip Any experience with the flag `os.O_TRUNC` before? Without it, the file isn't being overwritten completely. Is it an alternative to delete and recreate a file?

Reply:
I haven't used it in go, but it sounds like the truncate option most languages have. In https://docs.python.org/3/library/functions.html#open it's the default behavior of 'w' and one would have to use 'a' to append.

Reply:
Like we discussed in our 1:1, O_TRUNC is the default option for open file for many languages, but not golang. It allows to overwrite the file